src/ros_commo.py

```diff
@@ -376,15 +402,85 @@ def behavior_switch_callback(self, data):
 	# Data is a bit-flag that enables/disables publication of messages.
 	def behavior_control_callback(self, data):
 		self.control_mode = data.data
-
+	
+	def face_loc_cb(self, data):
+		#print "@@@@@@@@@@ In face cb"
+		self.lock.acquire()
+		try:
+			if len(self.sc_str_set)<1:
+				self.face_array=data.faces
+				self.sc_str_set="do"
+		finally:
+			self.lock.release()
+		#for face in data.faces:
+			#fac="(map-ato \"faces\" (NumberNode \""+str(face.id)+"\") "+str(face.point.x)+" "+str(face.point.y)+" "+str(face.point.z)+")"
+			#scheme_eval(self.atomspace,fac)
+
+	def thr(self):
+		self.atomspace = AtomSpace()#scheme_eval_as('(cog-atomspace)')
+		lpth="/home/mandeep/hr/opencog/ros-behavior-scripting/src/time-map.scm"
+		scheme_eval(self.atomspace, "(load \""+lpth+"\")")
+		#print scheme_eval(self.atomspace,"(+ 2 3)")
+		#scheme_eval(self.atomspace, "(map-ato \"faces\" (NumberNode \"2.0\") 1 2 3)")
+		rate=rospy.Rate(30)
+		while not rospy.is_shutdown():
+			rate.sleep()
+			self.lock.acquire()
+			try:
+				if len(self.sc_str_set)>0:
+					#print "faces: ", len(self.face_array)
+					for face in self.face_array:
+						#print "hello"
+						fac="(map-ato \"faces\" (NumberNode \""+str(face.id)+"\") "+str(face.point.x)+" "+str(face.point.y)+" "+str(face.point.z)+")"
+						scheme_eval(self.atomspace,fac)
+					self.sc_str_set=""
+
+				if len(self.sc_str_get)>0:
+					fid=self.sc_str_get
+					print "******look at face point: ", fid
+					fc="(look-at-face (NumberNode \""+fid+"\"))"
+					ptstr=scheme_eval(self.atomspace,fc)
+					if len(ptstr)<5:
+						print "XX Face Point Invalid XX"
+					else:
+						pts=ptstr.split()
+						trg = Target()
+						trg.x = float(pts[0])
+						trg.y = float(pts[1])
+						trg.z = float(pts[2])
+						self.turn_pub.publish(trg)
+						self.gaze_pub.publish(trg)
+					self.sc_str_get=""
+			finally:
+				self.lock.release()
+		
 	def __init__(self):
 
 		self.puta = PutAtoms()
 
 		# The below will hang until roscore is started!
 		rospy.init_node("OpenCog_Eva")
 		print("Starting OpenCog Behavior Node")
-
+		self.lock=threading.Lock();
+		self.sc_str_set=""
```

Comment:
why are you using python threads? They are a bad idea, in general-- python is only single-threaded, so that all happens is that everything will go more slowly.